### PR TITLE
Add HTML extension

### LINKED ISSUE #90: Support for non-standard macros \class and \cssId
Hi!

Support for the [TeX HTML extension from MathJax](http://docs.mathjax.org/en/latest/tex.html#html) in the output of KaTeX would be nice.

---

Excerpt from the docs:

The HTML extension gives you access to some HTML features like styles, classes, element ID’s and clickable links. It defines the following non-standard macros: 
- `\class{name}{math}`: Attaches the CSS class name to the output associated with math when it is included in the HTML page. This allows your CSS to style the element.
- `\cssId{id}{math}`: Attaches an id attribute with value id to the output associated with math when it is included in the HTML page. This allows your CSS to style the element, or your javascript to locate it on the page.

### LINKED ISSUE #2029: Support for \style{}
> Before requesting a feature, please search for [existing issues](https://github.com/KaTeX/KaTeX/issues).

**Is your feature request related to a problem? Please describe.**

When I am trying to render a latex - `\frac{\style{font-size:14px}R\style{font-size:14px}{\mathit=}\style{font-size:14px}\rho\style{font-size:14px}{\mathit\;}\style{font-size:14px}l}{\style{font-size:14px}A}` 
Screenshot 2019-07-10 at 4 20 27 PM
The latex is generated using Wiris's MathType Plugin.

**Link to or name of a (La)TeX package that provides the same feature:**

KaTeX's general goal is to render as LaTeX does.
When the same latex is entered here - http://www.wiris.com/editor/demo/en/developers - it renders as expected.

### PR DESCRIPTION
**Pull request by ylemkimon**
This PR adds four new commands that give access to some HTML features: `\htmlClass`, `\htmlId`, `\htmlStyle`, and `\htmlData`.

Fixes #90.
Fixes #2029.

### CODE REVIEW

File: src/functions/html.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import defineFunction, {ordargument} from "../defineFunction";
+import buildCommon from "../buildCommon";
+import {assertNodeType} from "../parseNode";
+
+import * as html from "../buildHTML";
+import * as mml from "../buildMathML";
+
+defineFunction({
+    type: "html",
+    names: ["\\htmlClass", "\\htmlId", "\\htmlStyle", "\\htmlData"],
```
Comment (ylemkimon) [inline src/functions/html.js]: In MathJax, they are `\class`, `\cssId`, and `\style`, respectively, (they don't have one that can set data attributes) but I think their names are too generic and we have/will have multiple renderers, so I named them with the prefix `html`.

Comment (ylemkimon) [reply src/functions/html.js]: `\htmlData` takes key/value pairs as the first argument and sets data attributes. Inspired by #1698.

File: src/functions/html.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        allowedInText: true,
+    },
+    handler: ({parser, funcName, token}, args) => {
+        const value = assertNodeType(args[0], "raw").string;
+        const body = args[1];
+
+        if (parser.settings.strict) {
+            parser.settings.reportNonstrict("htmlExtension",
+                "HTML extension is disabled on strict mode");
+        }
```
Comment (ylemkimon) [inline src/functions/html.js]: I'm not sure they are necessary, as the `trust` setting controls them.

File: src/functions/html.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    command: "\\htmlId",
+                    id: value,
+                };
+                break;
+            case "\\htmlStyle":
+                attributes.style = value;
+                trustContext = {
+                    command: "\\htmlStyle",
+                    style: value,
+                };
```
Comment (ylemkimon) [inline src/functions/html.js]: I can't get Flow to work if I use `funcName` as `command` in trustContexts.

File: src/functions/html.js (hunk 1 above)
Comment (kevinbarabash) [reply src/functions/html.js]: I like the prefixing.  Thanks for adding `\htmlData` as well.

File: src/functions/html.js (hunk 3 above)
Comment (kevinbarabash) [reply src/functions/html.js]: `funcName` is typed as `string`.  It looks like it might be possible to introduce another generic type param to `FunctionDefSpec` and related types to make that work.  I can poke around a bit.

**Comment by kevinbarabash** [comment]
@ylemkimon this is great!

**Comment by vd-man** [comment]
Does not work with fraction, eg:

\htmlId{id1}{a}+\frac{1}{\htmlId{id2}{b}}-c

id1 woks, but id2 does not.

**Comment by ylemkimon** [comment]
@vd-man Both `id1` and `id2` seem to work [here](https://deploy-preview-2082--katex.netlify.com/?data=%7B%22displayMode%22%3Atrue%2C%22leqno%22%3Afalse%2C%22fleqn%22%3Afalse%2C%22throwOnError%22%3Atrue%2C%22errorColor%22%3A%22%23cc0000%22%2C%22strict%22%3A%22ignore%22%2C%22trust%22%3Atrue%2C%22code%22%3A%22%5C%5ChtmlId%7Bid1%7D%7Ba%7D%2B%5C%5Cfrac%7B1%7D%7B%5C%5ChtmlId%7Bid2%7D%7Bb%7D%7D-c%22%7D). Could you give more detail? Thank you.

**Comment by ylemkimon** [comment]
This is ready for review!

**Comment by phlegmaticprogrammer** [comment]
Would it be possible to add some docs for what \htmlId, \htmlStyle, \htmlClass and in particular \htmlData actually do? How can I preview the docs? If I click on "Deploy preview ready!" I get to a preview which has no updated HTML section in the docs.

**Comment by kevinbarabash** [comment]
@phlegmaticprogrammer see https://deploy-preview-2082--katex.netlify.com/docs/next/supported.html#html

**Comment by phlegmaticprogrammer** [comment]
Now it works for me! This is great progress, although some CSS properties seem to be still under full KaTeX control, for example in 

`A*\htmlStyle{color: red; font-family: Arial, Helvetica, sans-serif; font-style: normal; font-size:50px;}{A}`

the color and font-size are applied, but the font-family and style are not. So a general \html would still be necessary I guess.

**Comment by kevinbarabash** [comment]
@phlegmaticprogrammer there's a ticket for adding `\html`, see #1596.

File: docs/issues.md [hunk 4]
```diff
@@ -22,6 +22,8 @@ title: Common Issues
 - MathJax defines `\color` to be like `\textcolor` by default; set KaTeX's
   `colorIsTextColor` option to `true` for this behavior.  KaTeX's default
   behavior matches MathJax with its `color.js` extension enabled.
+- Equivalents of MathJax `\class`, `\cssId`, and `\style` are `\htmlClass`,
+  `\htmlId`, and `\htmlStyle`, respectively, to avoid ambiguity.
```
Comment (kevinbarabash) [inline docs/issues.md]: Thanks for adding this.  If people want to use the MathJax variants they can set up some macros.

File: src/functions/html.js (hunk 3 above)
Comment (kevinbarabash) [reply src/functions/html.js]: I didn't have time to poke around.  This works.  This is something we can always revisit later.

File: contrib/render-a11y-string/render-a11y-string.js [hunk 5]
```diff
@@ -655,6 +655,10 @@ const handleObject = (
             break;
         }
 
+        case "html": {
+            throw new Error("KaTeX-a11y: html not implemented yet");
+        }
```
Comment (kevinbarabash) [inline contrib/render-a11y-string/render-a11y-string.js]: The `\html*` commands wrap other commands.  I think we could probably call `buildString` on `tree.body`.  Do you see any issues with that approach?

File: docs/support_table.md [hunk 6]
```diff
@@ -457,6 +457,10 @@ use `\ce` instead|
 |\hskip|$w\hskip1em i\hskip2em d$|`w\hskip1em i\hskip2em d`|
 |\hslash|$\hslash$||
 |\hspace|$s\hspace7ex k$|`s\hspace7ex k`|
+|\htmlClass|$\htmlClass{foo}{x}$|`\htmlClass{foo}{x}` Requires `trust` and `strict` [option](options.md)|
+|\htmlData|$\htmlData{foo=a, bar=b}{x}$|`\htmlData{foo=a, bar=b}{x}` Requires `trust` and `strict` [option](options.md)|
+|\htmlId|$\htmlId{bar}{x}$|`\htmlId{bar}{x}` Requires `trust` and `strict` [option](options.md)|
+|\htmlStyle|$\htmlStyle{color: red;}{x}$|`\htmlStyle{color: red;}{x}` Requires `trust` and `strict` [option](options.md)|
```
Comment (kevinbarabash) [inline docs/support_table.md]: This makes it sound like we need to set `trust: true` and `strict: true`.  Maybe reword it to say something like:
```
Must enable `trust` and disable `strict` [option](options.md)
```

File: src/functions/html.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                break;
+            case "\\htmlData": {
+                const data = value.split(",");
+                for (let i = 0; i < data.length; i++) {
+                    const keyVal = data[i].split("=");
+                    if (keyVal.length === 2) {
+                        attributes["data-" + keyVal[0].trim()] = keyVal[1].trim();
+                    }
```
Comment (kevinbarabash) [inline src/functions/html.js]: We should probably throw an error if `keyVal.length !== 2`.

**Comment by mariusGundersen** [comment]
This seems to be merged, but not released yet. What is the timeline for having this in a version available for use? I would really like to use this feature soon
